src/Ocelot.Provider.Consul/ConsulServiceBuilder.cs

```diff
@@ -0,0 +1,99 @@
+﻿using Ocelot.Infrastructure.Extensions;
+using Ocelot.Logging;
+using Ocelot.Provider.Consul.Interfaces;
+using Ocelot.Values;
+
+namespace Ocelot.Provider.Consul;
+
+public class ConsulServiceBuilder : IConsulServiceBuilder
+{
+    private readonly ConsulRegistryConfiguration _configuration;
+    private readonly IConsulClient _client;
+    private readonly IOcelotLogger _logger;
+
+    public ConsulServiceBuilder(
+        Func<ConsulRegistryConfiguration> configurationFactory,
+        IConsulClientFactory clientFactory,
+        IOcelotLoggerFactory loggerFactory)
+    {
+        _configuration = configurationFactory.Invoke();
+        _client = clientFactory.Get(_configuration);
+        _logger = loggerFactory.CreateLogger<ConsulServiceBuilder>();
+    }
+
+    public ConsulRegistryConfiguration Configuration => _configuration;
+    protected IConsulClient Client => _client;
+    protected IOcelotLogger Logger => _logger;
+
+    public virtual bool IsValid(ServiceEntry entry)
+    {
+        var address = entry.Service.Address;
+        return !string.IsNullOrEmpty(address)
+            && !address.Contains($"{Uri.UriSchemeHttp}://")
+            && !address.Contains($"{Uri.UriSchemeHttps}://")
+            && entry.Service.Port > 0;
+    }
+
+    public virtual IEnumerable<Service> BuildServices(ServiceEntry[] entries, Node[] nodes)
```

Comment:
Why the virtuals everywhere?

Reply:
### The quote 1
>In conclusion, customization hinges on **the virtual methods** of the `ConsulServiceBuilder` class, allowing developers to override as needed for highly customized Consul setups in their environments.

Thus, the concept is to permit developers to create a new class, inherit from the `ConsulServiceBuilder` class, and then simply override the necessary virtual methods to introduce new behavior.

### The quote 2
>  - The targeted virtual methods which solve the [954](https://github.com/ThreeMammals/Ocelot/issues/954) problem are:
>    - `protected virtual ServiceHostAndPort GetServiceHostAndPort(ServiceEntry entry, Node node)`
>    - `protected virtual string GetDownstreamHost(ServiceEntry entry, Node node)`

The author of issue #954 proposed reverting to the previous implementation that constructs a host from `ServiceEntry.Service.Address`.
To achieve this, a developer would need to override the method as follows:
```csharp
public class MyConsulServiceBuilder : ConsulServiceBuilder
{ 
    protected override string GetDownstreamHost(ServiceEntry entry, Node node) => entry.Service.Address;
}
```
Unfortunately, we cannot alter our current implementations as it would lead to breaking changes. It is preferable to provide developers with the flexibility to override as needed for their requirements.

Reply:
Ok, fine, but then, why not `AddConsul<T>` ?

Reply:
New method was added in commit https://github.com/ThreeMammals/Ocelot/pull/2067/commits/414d8dc3085d32028e4b9aebeee6c74845cb5d74 ✔️